src/services/uniswap.js

```diff
@@ -0,0 +1,72 @@
+import Ethereum from '../services/eth';
+const uni = require('@uniswap/sdk')
+const ethers = require('ethers')
+const proxyArtifact = require('../static/uniswap_v2_router_abi.json')
+const debug = require('debug')('router')
+
+const GAS_LIMIT = 1200000  //120000000
+
+export default class Uniswap {
+  constructor (network = 'kovan') {
+    // network defaults to KOVAN
+    const providerUrl = process.env.ETHEREUM_RPC_URL
+    this.network = process.env.UNISWAP_NETWORK
+    this.chainID = uni.ChainId.KOVAN
+    this.provider = new ethers.providers.JsonRpcProvider(providerUrl)
+    this.exchangeProxy = process.env.UNISWAP_PROXY
+
+    if (network === 'kovan') {
+      this.chainID = uni.ChainId.KOVAN
+    } else if (network === 'mainnet') {
+      this.chainID = uni.ChainId.MAINNET
+    } else {
+      throw Error(`Invalid network ${network}`)
+    }
+  }
+
+  async fetch_route(tokenIn, tokenOut){
+      var route, pair, pairOne, pairTwo
+      var tIn = await uni.Fetcher.fetchTokenData(this.chainID, tokenIn)
+      var tOut = await uni.Fetcher.fetchTokenData(this.chainID, tokenOut)
+
+      try {
+      pair = await uni.Fetcher.fetchPairData(tIn, tOut)
+      route = new uni.Route([pair], tIn, tOut)
+      }
+      catch(err){
+      console.log('Trying alternative/indirect route.')
+      pairOne = await uni.Fetcher.fetchPairData(tIn, uni.WETH[this.chainID])
+      pairTwo = await uni.Fetcher.fetchPairData(tOut, uni.WETH[this.chainID])
+      route = new uni.Route([pairOne, pairTwo], tIn, tOut)
+      }
+      return route
+  }
+
+  async swapExactIn (wallet, route, tokenAddress, amountIn, gasPrice) {
+    const tIn = await uni.Fetcher.fetchTokenData(this.chainID, tokenAddress)
+    const tokenAmountIn = new uni.TokenAmount(tIn, amountIn)
+    const result = uni.Router.swapCallParameters(
+          uni.Trade.exactIn(route, tokenAmountIn),
+          { ttl: 50, recipient: wallet.address, allowedSlippage: new uni.Percent('1', '100') }
```

Comment:
What would be better values for the *ttl* (time to live) and the "allowedSlippage"?

Reply:
`ttl`:Uniswap defines this as "how long the swap is valid until it expires, in seconds." I think we should make `ttl` an constant equal to 0 that is specified in `uniswap.js`.

`allowedSlippage`: I think we should make this a constant equal to 0 in `uniswap.js` as well. Since we set slippage buffer through the Hummingbot client, I don't think we need to add more slippage buffer here.

Reply:
There shouldn't be problem with setting allowedSloppage to 0.
However, I believe there would be a problem with setting ttl to 0. I believe the swap wouldn't be executed. I'd test to confirm this.

Reply:
Sorry that was a typo - for `ttl` I think a value like 60 seconds would be a good initial value.